sdk/include/opentelemetry/sdk/common/atomic_shared_ptr.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+/**
+ * A wrapper to provide atomic shared pointers.
+ *
+ * This wrapper relies on std::atomic for gcc 4.8 and C++20, while using
+ * specializations of std::atomic_store and std::atomic_load in all other
+ * instances.
+ */
+#if __cplusplus > 201703L
+template <class T>
+class AtomicSharedPtr
+{
+public:
+  explicit AtomicSharedPtr(std::shared_ptr<T> ptr) noexcept : ptr_{std::move(ptr)} {}
+
+  void store(const std::shared_ptr<T> &other) noexcept
+  {
+    ptr_.store(other, std::memory_order_release);
+  }
+
+  std::shared_ptr<T> load() const noexcept { return ptr_.load(std::memory_order_acquire); }
+
+private:
+  std::atomic<std::shared_ptr<T>> ptr_;
+};
+#elif (__GNUC__ == 4 && (__GNUC_MINOR__ >= 8))
+template <class T>
+class AtomicSharedPtr
+{
+public:
+  explicit AtomicSharedPtr(std::shared_ptr<T> ptr) noexcept
+      : ptr_{new std::shared_ptr<T>(std::move(ptr))}
+  {}
+
+  ~AtomicSharedPtr() noexcept { delete ptr_.load(std::memory_order_acquire); }
+
+  void store(const std::shared_ptr<T> &other) noexcept
+  {
+    ptr_.store(new std::shared_ptr<T>(other), std::memory_order_release);
```

Comment:
Won't this leak memory? ptr_ is already holding a heap allocated shared_ptr that doesn't get freed, no?

Reply:
You're right. I'll submit another PR with a fix.

Reply:
https://github.com/open-telemetry/opentelemetry-cpp/pull/71